containerd-shim-wws-v1/src/main.rs

```diff
@@ -0,0 +1,221 @@
+use chrono::{DateTime, Utc};
+use containerd_shim as shim;
+use containerd_shim_wasm::sandbox::error::Error;
+use containerd_shim_wasm::sandbox::instance::{EngineGetter, InstanceConfig};
+use containerd_shim_wasm::sandbox::oci;
+use containerd_shim_wasm::sandbox::{Instance, ShimCli};
+use log::{error, info};
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::mpsc::channel;
+use std::sync::mpsc::Sender;
+use std::sync::Arc;
+use std::sync::{Condvar, Mutex};
+use std::thread;
+use tokio::runtime::Runtime;
+use wws_config::Config;
+use wws_router::Routes;
+use wws_server::serve;
+
+/// URL to listen to in wws
+const WWS_ADDR: &str = "0.0.0.0";
+const WWS_PORT: u16 = 80;
+
+type ExitCode = Arc<(Mutex<Option<(u32, DateTime<Utc>)>>, Condvar)>;
+
+pub struct Workers {
+    exit_code: ExitCode,
+    id: String,
+    // TODO: Pass the stdio to wws so the logs are print from the pod
+    // stdin: String,
+    // stdout: String,
+    // stderr: String,
+    bundle: String,
+    shutdown_signal: Arc<(Mutex<bool>, Condvar)>,
+}
+
+pub fn prepare_module(bundle: String) -> Result<PathBuf, Error> {
+    info!("[wws] Preparing module");
+    let mut spec = oci::load(Path::new(&bundle).join("config.json").to_str().unwrap())
+        .expect("unable to load OCI bundle");
+
+    info!("[wws] Canonicalize roots");
+    spec.canonicalize_rootfs(&bundle)
+        .map_err(|err| Error::Others(format!("could not canonicalize rootfs: {err}")))?;
+
+    info!("[wws] Get root");
+    let working_dir = oci::get_root(&spec);
+    info!("[wws] loading project: {}", working_dir.display());
+
+    Ok(working_dir.clone())
+}
+
+/// Implement the "default" interface from runwasi
+impl Instance for Workers {
+    type E = ();
+    fn new(id: String, cfg: Option<&InstanceConfig<Self::E>>) -> Self {
+        info!("[wws] new instance");
+        let cfg = cfg.unwrap();
+
+        Workers {
+            exit_code: Arc::new((Mutex::new(None), Condvar::new())),
+            id,
+            // TODO: Pass the stdio to wws so the logs are print from the pod
+            // stdin: cfg.get_stdin().unwrap_or_default(),
+            // stdout: cfg.get_stdout().unwrap_or_default(),
+            // stderr: cfg.get_stderr().unwrap_or_default(),
+            bundle: cfg.get_bundle().unwrap_or_default(),
+            shutdown_signal: Arc::new((Mutex::new(false), Condvar::new())),
+        }
+    }
+
+    fn start(&self) -> Result<u32, Error> {
+        info!("[wws] Starting the wws shim");
+        let exit_code = self.exit_code.clone();
+        let shutdown_signal = self.shutdown_signal.clone();
+        let (tx, rx) = channel::<Result<(), Error>>();
+        let bundle = self.bundle.clone();
+
+        // TODO: Pass the stdio to wws so the logs are print from the pod
```

Comment:
It looks like the pod logs are not setup. Can you explain why this was not done?

Reply:
Currently, `wws` doesn't expose a way to modify the `stdio` in the WASI context. It inherits it by default from the environment. I'll work on exposing this capability too so the Wasm module properly prints the logs in the right place.

Reply:
I created this issue on `wws` to track the effort: https://github.com/vmware-labs/wasm-workers-server/issues/124

Reply:
I see. This makes more sense now.

Is there another way that a user will be able to see any kind of output from a `wws` workload? It's really tough to debug when something is not working without logs.

If there is no way right now, I think we can leave it as a high priority TODO.

Reply:
@devigned the `stderr` output is not used internally by `wws`. What do you think about using it for debugging purposes? I know it's not ideal, but at least we can provide a way to debug it :)

Reply:
That’s a great interview solution.

Reply:
Great @devigned! I will work on that 😄

Reply:
I created this task on the `wws` repo to track this effort: https://github.com/vmware-labs/wasm-workers-server/issues/125

Reply:
Done! I just merged the changes and updated the references to use the latest changes 😄